### ISSUE TITLE
CMake refactoring

### ISSUE THREAD

**Issue by rasolca** [eth-cscs/DLA-interface#84]

**Comment by Iceman9** [comment]
What is the current state of this?
 
I installed DLA-interface on our cluster but I had to manually specify the locations of LAPACK, BLAS and SCALAPACK. The CMake scripts doesn't find them on their own.

And even then, when I specify the locations, CMake fails at looking for functions:
```-- Looking for dgemm_
-- Looking for dgemm_ - not found
CMake Error at cmake/FindLAPACK.cmake:58 (message):
  BLAS symbol not found with this configuration
Call Stack (most recent call first):
  CMakeLists.txt:87 (find_package)
```
If I comment the lines in which CMake looks for functions, DLA-interface compiles and all tests except `test_communicator_grid.cpp  test_communicator_manager.cpp` fail.

I also checked branch **cmake_refactoring** and get the same result.

**Comment by Iceman9** [comment]
During rebasing of https://github.com/eth-cscs/DLA-interface/pull/90 and https://github.com/eth-cscs/DLA-interface/pull/88 found an issue with:

- https://github.com/eth-cscs/DLA-interface/blob/ffc0d3c42211f273fc1ac4e7684dae04f21cb54d/cmake/DLAI_AddTargetWarnings.cmake#L23
Generator COMPILE_LANGUAGE takes one parameter or zero (). The following line fails with specified minimum CMake 3.14 (https://cmake.org/cmake/help/v3.14/manual/cmake-generator-expressions.7.html). This can be fixed by raising the minimum CMake required version to 3.15 or specifying generator separate for both languages.

**Comment by albestro** [comment]
> During rebasing of #90 and #88 found an issue with:
> 
>     * https://github.com/eth-cscs/DLA-interface/blob/ffc0d3c42211f273fc1ac4e7684dae04f21cb54d/cmake/DLAI_AddTargetWarnings.cmake#L23
>       
>       Generator COMPILE_LANGUAGE takes one parameter or zero (). The following line fails with specified minimum CMake 3.14 (https://cmake.org/cmake/help/v3.14/manual/cmake-generator-expressions.7.html). This can be fixed by raising the minimum CMake required version to 3.15 or writing generators separate for both languages.

Good catch! I would go for upgrading our minimum requirement to cmake 3.15. @rasolca is it ok for you?
